packages/syft/src/syft/core/node/common/action/smpc_action.py

```diff
@@ -0,0 +1,258 @@
+
+    result_id = UID(UUID(bytes=generator.bytes(16)))
+    other = node.store[other_id].data
+
+    actions = []
+    if isinstance(other, ShareTensor):
+        # All parties should add the other share if empty list
+        actions.append(
+            SMPCAction(
+                "mpc_add",
+                self_id=self_id,
+                args_id=[other_id],
+                kwargs_id={},
+                ranks_to_run_action=[],
+                result_id=result_id,
+                address=node.address,
+            )
+        )
+    else:
+        # Only rank 0 (the first party) would add that public value
+        actions.append(
+            SMPCAction(
+                "mpc_add",
+                self_id=self_id,
+                args_id=[other_id],
+                kwargs_id={},
+                ranks_to_run_action=[0],
+                result_id=result_id,
+                address=node.address,
+            )
+        )
+
+    return actions
+
+
+def smpc_mul(self_id, other_id, seed, node):
+    generator = np.random.default_rng(seed)
+
+    for _ in range(MAP_FUNC_TO_NR_GENERATOR_INVOKES["__mul__"]):
+        generator.bytes(16)
+
+    result_id = UID(UUID(bytes=generator.bytes(16)))
+    other = node.store[other_id].data
+
+    actions = []
+    if isinstance(other, ShareTensor):
+        raise ValueError("Not yet implemented Private Multiplication")
+    else:
+        # All ranks should multiply by that public value
+        actions.append(
+            SMPCAction(
+                "mpc_mul",
+                self_id=self_id,
+                args_id=[other_id],
+                kwargs_id={},
+                ranks_to_run_action=[],
+                result_id=result_id,
+                address=node.address,
+            )
+        )
+
+    return actions
+
+
+MAP_FUNC_TO_ACTION = {"__add__": smpc_add, "__mul__": smpc_mul}
+
+
+_MAP_ACTION_TO_FUNCTION = {
+    "mpc_add": operator.add,
+    "mpc_mul": operator.mul,
+}
+
+
+@bind_protobuf
+class SMPCAction(ImmediateActionWithoutReply):
+    def __init__(
+        self,
+        name_action,
+        self_id,
+        args_id,
+        kwargs_id,
+        result_id,
+        address: Address,
+        ranks_to_run_action: Optional[List[int]] = None,
+        msg_id: Optional[UID] = None,
+    ):
+        self.name_action = name_action
+        self.self_id = self_id
+        self.args_id = args_id
+        self.kwargs_id = kwargs_id
+        self.id_at_location = result_id
+        self.ranks_to_run_action = ranks_to_run_action
+        self.address = address
+        self.msg_id = msg_id
+        super().__init__(address=address, msg_id=msg_id)
+
+    @staticmethod
+    def filter_actions_after_rank(rank, actions):
+        return [
+            action
+            for action in actions
+            if action.ranks_to_run_action == [] or rank in action.ranks_to_run_action
+        ]
+
+    def execute_action(self, node: AbstractNode, verify_key: VerifyKey) -> None:
```

Comment:
I'm surprised you needed to re-write all this logic. Could you not have simply called the node service which can execute actions?

Reply:
I will check this out :+1:

Reply:
I am not sure in this case what the service would be -- I reach this point after:
```RunClassMethodAction``` (here I get all the actions - SMPCActions for the current node) --> I send them to SMPC Service and they execute it (and enter here)